feat(shared-data): Add "$otSharedSchema" property to labware schema 3

SyntaxColoring opened the pull request:
## Background

Labware definitions currently do not have any way of self-identifying as labware definitions. They are just JSON files, with a .json file extension, and they do not contain any field that explicitly identifies them as labware definitions. So the system has to rely on heuristics, which is a little gross:

https://github.com/Opentrons/opentrons/blob/19a94a630dc67d205495750c44d2d56b9bf8fd39/api/src/opentrons/protocol_reader/file_identifier.py#L183-L185

This fixes that in the yet-to-be-released labware schema 3.

Closes EXEC-1156.

## Test Plan and Hands on Testing

~~Our automated tests do not currently validate our labware definitions against schema 3. That's an accident and we're fixing it in #17425. Meanwhile, I've verified locally that the updated definitions conform to the updated schema.~~ We're covered by automated tests now.

## Changelog

* Add an `"$otSharedSchema": "#/labware/schemas/3"` top-level property to labware schema 3. This follows the pattern set by [our JSON protocol schema](https://github.com/Opentrons/opentrons/blob/890e550116684fc302c308e77f87b4e35548754f/shared-data/protocol/schemas/8.json#L24-L27) and others.
* Update Python and TS bindings accordingly.
* Update labware definitions accordingly.

## Review requests

See comments.

## Risk assessment

At least medium because of Pydantic round-trip lossiness—see comments.

High if I'm also right about schema 3 already being released into the wild—see comments.

SyntaxColoring committed:
Add $otSharedSchema required property to labware schema.

SyntaxColoring committed:
Update labware definitions.

SyntaxColoring committed:
Update labware test fixtures.

SyntaxColoring committed:
Update Python and TypeScript bindings.

github-actions[bot] committed:
fix(analyses-snapshot-testing): heal identify_labware_defs snapshots (#17423)

This PR was requested on the PR
https://github.com/Opentrons/opentrons/pull/17421

Co-authored-by: SyntaxColoring <3236864+SyntaxColoring@users.noreply.github.com>

SyntaxColoring commented on analyses-snapshot-testing/tests/__snapshots__/analyses_snapshot_test/test_analysis_snapshot[2b866b03f3][Flex_S_v2_21_P1000_96_GRIP_HS_MB_TC_TM_Smoke].json:
```diff
@@ -1240,6 +1240,7 @@
       },
       "result": {
         "definition": {
+          "$otSharedSchema": "#/labware/schemas/3",
```
It's surprising to me that any analysis snapshots need to change. These are for PAPIv2.21 protocols. PAPIv2.21 
 was released in robot software v8.2.0. **Are schema 3 labware already being used in the wild?** Because my understanding was that schema 3 was not ready for that.

It's a bummer if that's the case because it means it'll be difficult to make breaking schema changes. And if we can't make any breaking schema changes then we may as well have stayed on schema 2 in the first place.

sfoster1 replied:
Interesting. It looks like the answer to this is that these labware have only a schema v3 definition:
- evotips_flex_96_tiprack_adapter
- evotips_opentrons_96_labware
- opentrons_tough_pcr_auto_sealing_lid
- protocol_engine_lid_stack_object
- opentrons_flex_tiprack_lid

and what's in these analyses is the thermocycler lid. these haven't shipped yet, so as long as we make sure there's v2 definitions in 8.3 before it ships we're good.

SyntaxColoring replied:
For posterity: the above is EXEC-1179 + EXEC-1180.

SyntaxColoring commented on shared-data/labware/schemas/3.json:
```diff
@@ -318,6 +319,10 @@
     "groups"
   ],
   "properties": {
+    "$otSharedSchema": {
+      "description": "The path to a valid Opentrons shared schema relative to the shared-data directory, without its extension.",
+      "enum": ["#/labware/schemas/3"]
+    },
```
Do we like `$otSharedSchema` for this, or do we want to try use something like `$id` instead? `$id` is [more standard](https://json-schema.org/understanding-json-schema/basics#declaring-a-unique-identifier) but it would be new to us.

sfoster1 replied:
I would say just stick with `$otSharedSchema` if we're not doing a full uri ref to a schema

SyntaxColoring replied:
I was misunderstanding the `$id` property. The `$id` property appears in schemas, not in documents conforming to those schemas.

SyntaxColoring committed:
Grammar.

SyntaxColoring commented on shared-data/python/opentrons_shared_data/labware/labware_definition.py:
```diff
@@ -684,6 +684,18 @@ class InnerWellGeometry(BaseModel):
 
 
 class LabwareDefinition(BaseModel):
+    model_config = {
+        # `"extra": "allow"` gives us lossless preservation of the $otSharedSchema field
+        # (which should always be omitted in schema 2 and always be present in schema 3)
+        # across parse/serialize round trips. Pydantic doesn't seem to have a good way
+        # of doing that when the $otSharedSchema field is declared explicitly on this
+        # model.
+        #
+        # Splitting this model into separate ones for schemas 2 and 3 would also solve this.
+        "extra": "allow"
+    }
+
+    # $otSharedSchema deliberately omitted for now. See `"extra": "allow"` in model_config.
     schemaVersion: Literal[1, 2, 3] = Field(
         ..., description="Which schema version a labware is using"
     )
```
Is this a reasonable way to avoid Pydantic hazards, at least for now?

sfoster1 replied:
Should be, yeah

SyntaxColoring replied:
Also, we're splitting the models in EXEC-1206, and this should get a bit better after we do that.

SyntaxColoring commented:
Re-drafting while https://github.com/Opentrons/opentrons/pull/17421#discussion_r1942986614 is worked out.

SyntaxColoring committed:
Revert "fix(analyses-snapshot-testing): heal identify_labware_defs snapshots (#17423)"

This reverts commit 8e2e01fa9ee5b4134ee677bdac5b505da0011364.

SyntaxColoring committed:
Merge branch 'edge' into identify_labware_defs

SyntaxColoring committed:
Update evotips_flex_96_tiprack_adapter and evotips_opentrons_96_labware.

SyntaxColoring committed:
Merge branch 'edge' into identify_labware_defs

SyntaxColoring committed:
Merge branch 'edge' into identify_labware_defs

SyntaxColoring committed:
Merge branch 'edge' into identify_labware_defs

Resolve conflicts in shared-data/python/opentrons_shared_data/labware/types.py

SyntaxColoring committed:
Include $otSharedSchema.

It turns out we do need this for the validation tests to pass--otherwise typeguard complains about extra properties.

sfoster1 reviewed (approved):
Looks good to me, and it's a good idea. We have not released schema 3 labware.

SyntaxColoring committed:
Merge branch 'edge' into identify_labware_defs

Resolve conflicts in shared-data/python/opentrons_shared_data/labware/labware_definition.py
